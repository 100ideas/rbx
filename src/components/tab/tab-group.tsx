import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import { forwardRefAs, Generic, HelpersProps } from "../../base";
import { Prefer } from "../../types";
import { tuple } from "../../utils";

export const TAB_GROUP_DEFAULTS = {
  alignments: tuple("centered", "right"),
  sizes: tuple("small", "medium", "large"),
  types: tuple("boxed", "toggle", "toggle-rounded"),
};

export interface TabGroupVariablesOverrides {}

export interface TabGroupVariablesDefaults {
  alignments: (typeof TAB_GROUP_DEFAULTS["alignments"])[number];
  sizes: (typeof TAB_GROUP_DEFAULTS["sizes"])[number];
  types: (typeof TAB_GROUP_DEFAULTS["types"])[number];
}

export type TabGroupVariables = Prefer<
  TabGroupVariablesOverrides,
  TabGroupVariablesDefaults
>;

export type TabGroupModifierProps = Partial<{
  align: TabGroupVariables["alignments"];
  fullwidth: boolean;
  size: TabGroupVariables["sizes"];
  /** * This is called style on Bulma documentation */
  type: TabGroupVariables["types"]; // tslint:disable-line:no-reserved-keywords
}>;

export type TabGroupProps = HelpersProps & TabGroupModifierProps;

const propTypes = {
  align: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  fullwidth: PropTypes.bool,
  size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  type: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

export const TabGroup = Object.assign(
  forwardRefAs<TabGroupProps, "div">(
    ({ align, children, className, fullwidth, size, type, ...rest }, ref) => (
      <Generic
        className={classNames(
          "tabs",
          {
            [`is-${align}`]: align,
            [`is-${size}`]: size,
            "is-fullwidth": fullwidth,
            "is-toggle": type === "toggle" || type === "toggle-rounded",
            "is-toggle-rounded": type === "toggle-rounded",
            [`is-${type}`]: type,
          },
          className,
        )}
        children={<ul>{children}</ul>}
        ref={ref}
        {...rest}
      />
    ),
    { as: "div" },
  ),
  { propTypes },
);
