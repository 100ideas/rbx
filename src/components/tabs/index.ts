import "./tabs.module.sass";

export { Tabs } from "./tabs";
