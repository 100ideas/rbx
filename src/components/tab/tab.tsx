import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import { forwardRefAs, Generic, HelpersProps } from "../../base";
import { TabGroup } from "./tab-group";

export type TabModifierProps = Partial<{
  active: boolean;
}>;

export type TabProps = HelpersProps & TabModifierProps;

const propTypes = {
  active: PropTypes.bool,
};

export const Tab = Object.assign(
  forwardRefAs<TabProps, "a">(
    ({ active, ...rest }, ref) => (
      <li className={classNames({ "is-active": active })}>
        <Generic ref={ref} {...rest} />
      </li>
    ),
    { as: "a" },
  ),
  {
    Group: TabGroup,
    propTypes,
  },
);
