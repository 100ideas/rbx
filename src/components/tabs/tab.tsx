import { cx } from "emotion";
import React from "react";

import { Element } from "@/components/element";
import { renderAsExoticComponent } from "@/components/render-as-exotic-component";
import { ModifierProps } from "@/modifiers";

export type TabsTabModifierProps = Partial<{
  active: boolean;
  children: React.ReactNode;
  style: React.CSSProperties;
}>;

export type TabsTabProps = ModifierProps & TabsTabModifierProps;

export const Tab = renderAsExoticComponent<TabsTabModifierProps, "a">(
  ({ children, className, style, active, ...props }, ref) => (
    <li
      ref={ref}
      style={style}
      className={cx(className, {
        "is-active": active,
      })}
    >
      <Element {...props}>{children}</Element>
    </li>
  ),
  "a",
);
Tab.defaultProps = Object.assign(
  {
    active: false,
    children: null,
  },
  Tab.defaultProps,
);
