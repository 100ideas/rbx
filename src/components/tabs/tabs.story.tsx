import { boolean, select } from "@storybook/addon-knobs";
import { storiesOf } from "@storybook/react";
import React from "react";

import { Tabs } from "@/components/tabs";
import { TabsProps } from "@/components/tabs/tabs";

// https://github.com/storybooks/storybook/issues/4865
const alignSelectOptions = ({
  Default: null,
  centered: "centered",
  right: "right",
} as unknown) as { [k: string]: string };

const makeAlignSelect = () => select("Align", alignSelectOptions, "Default");

storiesOf("Tabs", module)
  .addDecorator(story => <div style={{ margin: 10 }}>{story()}</div>)
  .add("Default", () => (
    <Tabs
      type={select(
        "Tab type",
        {
          boxed: "boxed",
          toggle: "toggle",
          "toggle-rounded": "toggle-rounded",
        },
        "boxed",
      )}
      fullwidth={boolean("Full width", false)}
      align={makeAlignSelect() as TabsProps["align"]}
    >
      <Tabs.Tab active>Test</Tabs.Tab>
      <Tabs.Tab>Test</Tabs.Tab>
      <Tabs.Tab>Test</Tabs.Tab>
      <Tabs.Tab>Test</Tabs.Tab>
    </Tabs>
  ));
