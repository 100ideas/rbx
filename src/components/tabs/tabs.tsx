import { cx } from "emotion";
import React from "react";

import { asExoticComponent } from "@/components/exotic";
import { ModifierProps, modify } from "@/modifiers";
import { Tab } from "./tab";

export type TabsModifierProps = Partial<{
  align: "centered" | "right";
  fullwidth: boolean;
  size: "small" | "medium" | "large";
  /** This is called style on Bulma documentation */
  type: "toggle" | "boxed" | "toggle-rounded";
}>;

export type TabsProps = ModifierProps & TabsModifierProps;

export const Tabs = Object.assign(
  asExoticComponent<TabsProps, "div">((props, ref) => {
    const { align, as, children, fullwidth, size, type, ...rest } = modify(
      props,
    );
    rest.className = cx("tabs", rest.className, {
      [`is-${align}`]: align,
      [`is-${size}`]: size,
      "is-fullwidth": fullwidth,
      // todo: Bulma 0.6.2 is not releaset ATM
      "is-toggle": type === "toggle-rounded",
      [`is-${type}`]: type,
    });
    return React.createElement(as!, {
      children: <ul>{children}</ul>,
      ref,
      ...rest,
    });
  }, "div"),
  { Tab },
);

Tabs.defaultProps = Object.assign({ fullwidth: false }, Tabs.defaultProps);
