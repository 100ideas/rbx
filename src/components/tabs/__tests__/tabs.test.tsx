import React from "react";

import { Tab } from "../tab";
import { Tabs, TABS_ALIGNMENTS, TABS_SIZES, TABS_TYPES } from "../tabs";

import {
  hasProperties,
  makeNodeFactory,
  makeShallowWrapper,
  testForwardRefAsExoticComponentIntegration,
  testThemeIntegration,
  validateBoolPropType,
  validateOneOfPropType,
} from "../../../__tests__/testing";

const COMPONENT = Tabs;
const COMPONENT_NAME = "Tabs";
const DEFAULT_ELEMENT = "div";
const BULMA_CLASS_NAME = "tabs";

const makeNode = makeNodeFactory(COMPONENT);

describe(`${COMPONENT_NAME} component`, () => {
  hasProperties(COMPONENT, {
    Tab,
    defaultProps: { as: DEFAULT_ELEMENT },
  });

  testForwardRefAsExoticComponentIntegration(
    makeNode,
    makeShallowWrapper,
    DEFAULT_ELEMENT,
    BULMA_CLASS_NAME,
  );

  testThemeIntegration(makeNode, makeShallowWrapper);

  describe("props", () => {
    const { propTypes } = COMPONENT;

    describe("align", () => {
      validateOneOfPropType(propTypes, "align", TABS_ALIGNMENTS);

      TABS_ALIGNMENTS.map(align =>
        it(`should be ${align}`, () => {
          const node = makeNode({ align });
          const wrapper = makeShallowWrapper(node);
          expect(wrapper.hasClass(`is-${align}`)).toBe(true);
        }),
      );
    });

    describe("children", () => {
      it("should wrap children in an ul element", () => {
        const node = makeNode({ children: <div className="foo" /> });
        const wrapper = makeShallowWrapper(node);
        const children = wrapper.children();
        expect(children.is("ul")).toBe(true);
        expect(children.children().is("div")).toBe(true);
        expect(children.children().hasClass("foo")).toBe(true);
      });
    });

    describe("fullwidth", () => {
      validateBoolPropType(propTypes, "fullwidth");

      [false, true].map(fullwidth =>
        it(`should ${fullwidth ? "" : "not "}be fullwidth`, () => {
          const node = makeNode({ fullwidth });
          const wrapper = makeShallowWrapper(node);
          expect(wrapper.hasClass("is-fullwidth")).toBe(fullwidth);
        }),
      );
    });

    describe("size", () => {
      validateOneOfPropType(propTypes, "size", TABS_SIZES);

      TABS_SIZES.map(size =>
        it(`should be ${size}`, () => {
          const node = makeNode({ size });
          const wrapper = makeShallowWrapper(node);
          expect(wrapper.hasClass(`is-${size}`)).toBe(true);
        }),
      );
    });

    describe("type", () => {
      validateOneOfPropType(propTypes, "type", TABS_TYPES);

      TABS_TYPES.map(type =>
        it(`should be ${type}`, () => {
          const node = makeNode({ type });
          const wrapper = makeShallowWrapper(node);
          expect(wrapper.hasClass(`is-${type}`)).toBe(true);
        }),
      );
    });
  });
});

// import Enzyme from "enzyme";
// import React from "react";

// import { Tab } from "../tab";
// import { Tabs, TABS_ALIGNMENTS, TABS_SIZES, TABS_TYPES } from "../tabs";

// import {
//   hasProperties,
//   testGenericPropTypes,
//   validateBoolPropType,
//   validateOneOfPropType,
// } from "../../../__tests__/testing";

// describe("Tabs component", () => {
//   hasProperties(Tabs, {
//     Tab,
//     defaultProps: { as: "div" },
//   });

//   it("should render as the default element", () => {
//     const wrapper = Enzyme.shallow(<Tabs />);
//     expect(wrapper.is("div")).toBe(true);
//   });

//   it("should render as a custom component", () => {
//     const as = "span";
//     const wrapper = Enzyme.shallow(<Tabs as={as} />);
//     expect(wrapper.is(as)).toBe(true);
//   });

//   it("should forward ref", () => {
//     const ref = React.createRef<HTMLDivElement>();
//     // Enzyme owns outer ref: https://github.com/airbnb/enzyme/issues/1852
//     const wrapper = Enzyme.mount(
//       <div>
//         <Tabs ref={ref} />
//       </div>,
//     );
//     try {
//       expect(ref.current).toBe(wrapper.find(".tabs").instance());
//     } finally {
//       wrapper.unmount();
//     }
//   });

//   it("should have bulma className", () => {
//     const wrapper = Enzyme.shallow(<Tabs />);
//     expect(wrapper.hasClass("tabs")).toBe(true);
//   });

//   it("should preserve custom className", () => {
//     const className = "foo";
//     const wrapper = Enzyme.shallow(<Tabs className={className} />);
//     expect(wrapper.hasClass(className)).toBe(true);
//   });

//   TABS_ALIGNMENTS.map(align =>
//     it(`should be aligned ${align}`, () => {
//       const wrapper = Enzyme.shallow(<Tabs align={align} />);
//       expect(wrapper.hasClass(`is-${align}`)).toBe(true);
//     }),
//   );

//   TABS_SIZES.map(size =>
//     it(`should be size ${size}`, () => {
//       const wrapper = Enzyme.shallow(<Tabs size={size} />);
//       expect(wrapper.hasClass(`is-${size}`)).toBe(true);
//     }),
//   );

//   TABS_TYPES.map(type =>
//     it(`should be type ${type}`, () => {
//       const wrapper = Enzyme.shallow(<Tabs type={type} />);
//       expect(wrapper.hasClass(`is-${type}`)).toBe(true);
//       expect(wrapper.hasClass("is-toggle-rounded")).toBe(
//         type === "toggle-rounded",
//       );
//     }),
//   );

//   [false, true].map(fullwidth =>
//     it(`should ${fullwidth ? "" : "not "}be fullwidth`, () => {
//       const wrapper = Enzyme.shallow(<Tabs fullwidth={fullwidth} />);
//       expect(wrapper.hasClass("is-fullwidth")).toBe(fullwidth);
//     }),
//   );

//   describe("propTypes", () => {
//     const { propTypes } = Tabs;
//     testGenericPropTypes(propTypes);
//     validateOneOfPropType(propTypes, "align", TABS_ALIGNMENTS);
//     validateBoolPropType(propTypes, "fullwidth");
//     validateOneOfPropType(propTypes, "size", TABS_SIZES);
//     validateOneOfPropType(propTypes, "type", TABS_TYPES);
//   });
// });
