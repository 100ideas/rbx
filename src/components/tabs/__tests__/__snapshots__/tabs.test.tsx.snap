// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component should Exist 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "Tab": Object {
    "$$typeof": Symbol(react.forward_ref),
    "defaultProps": Object {
      "active": false,
      "as": "a",
    },
    "render": [Function],
  },
  "defaultProps": Object {
    "as": "div",
    "fullwidth": false,
  },
  "render": [Function],
}
`;

exports[`Tabs component should accept a react Element as renderAs prop 1`] = `
<p
  className="tabs"
>
  Custom
  <ul>
    This should be a p element
  </ul>
</p>
`;

exports[`Tabs component should concat Bulma class with classes in props 1`] = `
<div
  className="tabs other-class test"
>
  <ul>
    Facebook
  </ul>
</div>
`;

exports[`Tabs component should expose Tab 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "defaultProps": Object {
    "active": false,
    "as": "a",
  },
  "render": [Function],
}
`;

exports[`Tabs component should have Tabs classname 1`] = `
<div
  className="tabs"
>
  <ul>
    Facebook
  </ul>
</div>
`;

exports[`Tabs component should have custom inline styles 1`] = `
<section
  className="tabs"
  style={
    Object {
      "width": 200,
      "zIndex": 1,
    }
  }
>
  <ul>
    This should be a section with custom styles
  </ul>
</section>
`;

exports[`Tabs component should render as an html section 1`] = `
<section
  className="tabs"
>
  <ul>
    This should be a section
  </ul>
</section>
`;

exports[`Tabs component should render propertly Inside Tabs 1`] = `
<div
  className="tabs"
>
  <ul>
    <li
      className=""
    >
      <a>
        Tab 1
      </a>
    </li>
    <li
      className=""
    >
      <a>
        Tab 2
      </a>
    </li>
    <li
      className=""
    >
      <a>
        Tab 3
      </a>
    </li>
  </ul>
</div>
`;
